memory_unsafe.go

```diff
@@ -0,0 +1,303 @@
+// the Go heap. The function undoes any mappings and discards the span's
+// contents.
+//
+// Used as a finalizer in [newMemory], split off into a separate function for
+// testing and to avoid accidentally closing over the unsafe.Pointer to the
+// memory region, which would cause a cyclical reference.
+//
+// The resulting function panics if the mmap operation returns an error, since
+// it would mean the integrity of the Go heap is compromised.
+func unmap(size int) func(*byte) {
+	return func(a *byte) {
+		// Create another mapping at the same address to undo the original mapping.
+		// This will cause the kernel to repair the slab since we're using the same
+		// protection mode and flags as the original mapping for the Go heap.
+		//
+		// Address           Kbytes     RSS   Dirty Mode  Mapping
+		// 000000c000000000    4096     884     884 rw--- [ anon ]
+		//
+		// Using munmap here would leave an unmapped hole in the heap, compromising
+		// its integrity.
+		//
+		// MmapPtr allocates another unsafe.Pointer at the same address. Even though
+		// we discard it here, it may temporarily resurrect the backing array and
+		// delay its collection to the next GC cycle.
+		_, err := unix.MmapPtr(-1, 0, unsafe.Pointer(a), uintptr(size),
+			unix.PROT_READ|unix.PROT_WRITE,
+			unix.MAP_PRIVATE|unix.MAP_FIXED|unix.MAP_ANON)
+		if err != nil {
+			panic(fmt.Errorf("undoing bpf map memory mapping: %w", err))
+		}
+	}
+}
+
+// checkUnsafeMemory ensures value T can be accessed in mm at offset off.
+//
+// The comparable constraint narrows down the set of eligible types to exclude
+// slices, maps and functions. These complex types cannot be mapped to memory
+// directly.
+func checkUnsafeMemory[T comparable](mm *Memory, off uint64) error {
+	if mm.b == nil {
+		return fmt.Errorf("memory-mapped region is nil")
+	}
+	if mm.ro {
+		return ErrReadOnly
+	}
+	if !mm.heap {
+		return fmt.Errorf("memory region is not heap-mapped, see CollectionOptions.UnsafeVariableExperiment: %w", ErrNotSupported)
+	}
+
+	t := reflect.TypeFor[T]()
+	if err := checkType(t.String(), t); err != nil {
+		return err
+	}
+
+	size := t.Size()
+	if size == 0 {
+		return fmt.Errorf("zero-sized type %s: %w", t, ErrInvalidType)
+	}
+
+	align := internal.Align(off, uint64(t.Align()))
+	if off != align {
+		return fmt.Errorf("unaligned access of memory-mapped region: %d-byte aligned read at offset %d aligns to %d", t.Align(), off, align)
+	}
+
+	vs, bs := uint64(size), uint64(len(mm.b))
+	if off+vs > bs {
+		return fmt.Errorf("%d-byte value at offset %d exceeds mmap size of %d bytes", vs, off, bs)
+	}
+
+	return nil
+}
+
+// checkType recursively checks if the given type is supported for memory
+// mapping. Only fixed-size types are supported: bools, floats, (u)int[8-64],
+// arrays, and structs containing them.
+//
+// Doesn't check for loops since it rejects pointers. Should that ever change, a
+// visited set would be needed.
+func checkType(name string, t reflect.Type) error {
```

Comment:
I have a hunch that you'll have to allow `uintptr` because that is the natural way to represent a C pointer in Go. That only works as long as user space and kernel space are both 64 bit of course xD

Reply:
Good point, we can allow uintptr if it's 8 bytes.

Reminder: we still treat btf.Pointer as 64 bits https://github.com/cilium/ebpf/issues/963.